language/vm/src/proptest_types/types.rs

```diff
@@ -0,0 +1,226 @@
+        let idx = max(self.module_idx.index(module_len) as TableIndex, 1);
+        let mut type_parameters = vec![];
+        for type_param in self.type_parameters {
+            type_parameters.push(type_param.materialize());
+        }
+        StructHandle {
+            module: ModuleHandleIndex(idx as TableIndex),
+            name: IdentifierIndex(self.name_idx.index(identifiers_len) as TableIndex),
+            is_nominal_resource: self.is_nominal_resource,
+            type_parameters,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct StructDefinitionGen {
+    name_idx: PropIndex,
+    is_nominal_resource: bool,
+    type_parameters: Vec<KindGen>,
+    is_public: bool,
+    field_defs: Option<Vec<FieldDefinitionGen>>,
+}
+
+impl StructDefinitionGen {
+    pub fn strategy(
+        field_count: impl Into<SizeRange>,
+        kind_count: impl Into<SizeRange>,
+    ) -> impl Strategy<Value = Self> {
+        (
+            any::<PropIndex>(),
+            any::<bool>(),
+            vec(KindGen::strategy(), kind_count),
+            any::<bool>(),
+            option::of(vec(FieldDefinitionGen::strategy(), field_count)),
+        )
+            .prop_map(
+                |(name_idx, is_nominal_resource, type_parameters, is_public, field_defs)| Self {
+                    name_idx,
+                    is_nominal_resource,
+                    type_parameters,
+                    is_public,
+                    field_defs,
+                },
+            )
+    }
+
+    pub fn materialize(
+        self,
+        state: &mut StDefnMaterializeState,
+    ) -> (Option<StructDefinition>, usize) {
+        let mut field_names = HashSet::new();
+        let mut fields = vec![];
+        match self.field_defs {
+            None => (),
+            Some(field_defs_gen) => {
+                for fd_gen in field_defs_gen {
+                    let field = fd_gen.materialize(state);
+                    if field_names.insert(field.name) {
+                        fields.push(field);
+                    }
+                }
+            }
+        };
+        let is_nominal_resource = if fields.is_empty() {
+            self.is_nominal_resource
+        } else {
+            self.is_nominal_resource
+                || fields.iter().any(|field| {
+                    let field_sig = &field.signature.0;
+                    state.contains_nominal_resource(field_sig)
+                })
+        };
+        let handle = StructHandle {
+            // 0 represents the current module
+            module: ModuleHandleIndex(0),
+            name: IdentifierIndex(self.name_idx.index(state.identifiers_len) as TableIndex),
+            is_nominal_resource,
+            type_parameters: self
+                .type_parameters
+                .into_iter()
+                .map(|kind| kind.materialize())
+                .collect(),
+        };
+        match state.add_struct_handle(handle) {
+            Some(struct_handle) => {
```

Comment:
Style consideration

```
Some(struct_handle) if fields.is_empty() => ...,
Some(struct_handle) => ...,
```

I like trying to prevent nesting if possible, I just find it easier to read. But I could see it also becoming more convoluted if you repeat too many patterns

Reply:
I find the nesting much more readable so I'll live it as is.
Of course if someone wanted to change at a later point is fine by me